pkg/core/statesync/module.go

```diff
@@ -429,6 +508,108 @@ func (s *Module) AddMPTNodes(nodes [][]byte) error {
 	return nil
 }
 
+// AddContractStorageData adds a batch of key-value pairs for storage-based sync.
+func (s *Module) AddContractStorageData(kv []byte, syncHeight uint32, expectedRoot util.Uint256) error {
+	if s.mode == MPTNodes {
+		return errors.New("storage pair not expected in MPT-based mode")
+	}
+	oldStage := s.syncStage
+	s.lock.Lock()
+	defer func() {
+		if s.syncStage != oldStage {
+			s.notifyStageChanged()
+		}
+	}()
+	defer s.lock.Unlock()
+
+	if s.syncStage&headersSynced == 0 || s.syncStage&mptSynced != 0 {
+		return errors.New("contract storage items were not requested")
+	}
+	if syncHeight != s.syncPoint {
+		return fmt.Errorf("invalid sync height: got %d, expected %d", syncHeight, s.syncPoint)
+	}
+
+	var (
+		prefix  = TemporaryPrefix(s.dao.Version.StoragePrefix)
+		lastKey []byte
+		br      = io.NewBinReaderFromBuf(kv)
+		batch   = make(map[string][]byte)
+	)
+	for !errors.Is(br.Err, gio.EOF) {
+		key := br.ReadVarBytes()
+		if br.Err != nil {
+			if errors.Is(br.Err, gio.EOF) {
+				break
+			}
+			return fmt.Errorf("failed to read key: %w", br.Err)
+		}
+		value := br.ReadVarBytes()
+		if br.Err != nil {
+			return fmt.Errorf("failed to read value: %w", br.Err)
+		}
+		s.dao.Store.Put(append([]byte{byte(prefix)}, key...), value)
+		batch[string(key)] = value
+		lastKey = key
+	}
+
+	stagingStore := storage.NewMemCachedStore(s.dao.Store)
+	originalStore := s.localTrie.Store
+	s.localTrie.Store = stagingStore
+
+	if len(batch) > 0 {
+		mptBatch := mpt.MapToMPTBatch(batch)
+		if _, err := s.localTrie.PutBatch(mptBatch); err != nil {
+			s.localTrie.Store = originalStore // Reset store on error.
+			return fmt.Errorf("failed to apply MPT batch at height %d: %w", syncHeight, err)
+		}
+		s.localTrie.Flush(syncHeight)
+	}
+	if n, err := stagingStore.PersistSync(); err != nil {
+		s.localTrie.Store = originalStore
+		return fmt.Errorf("failed to persist staging store (%d items): %w", n, err)
+	}
+
+	s.localTrie.Store = originalStore
+	ckpt := checkpoint{
+		MPTRoot: s.localTrie.StateRoot(),
+		LastKey: lastKey,
+	}
+
+	bw := io.NewBufBinWriter()
+	ckpt.EncodeBinary(bw.BinWriter)
+	if bw.Err != nil {
+		return fmt.Errorf("failed to encode checkpoint metadata: %w", bw.Err)
+	}
+	s.dao.Store.Put([]byte{byte(storage.SYSTempStateCheckpoint)}, bw.Bytes())
+
+	computedRoot := s.localTrie.StateRoot()
+	if !computedRoot.Equals(expectedRoot) {
+		s.log.Warn("computed state root doesn't match expected",
+			zap.String("computed root", computedRoot.StringLE()),
+			zap.String("expected root", expectedRoot.StringLE()))
+		return nil
+	}
+
+	sr := &state.MPTRoot{
+		Index: s.syncPoint,
+		Root:  computedRoot,
+	}
+	s.stateMod.JumpToState(sr)
+	if n, err := s.dao.PersistSync(); err != nil {
+		return fmt.Errorf("failed to persist final state (%d items): %w", n, err)
+	}
+
+	s.dao.Store.Delete([]byte{byte(storage.SYSTempStateCheckpoint)})
```

Comment:
This should be a part of `jumpToStateInternal` because it should be an atomic change unified with statejump.

Reply:
> This should be a part of jumpToStateInternal

Or may be not, because from another hand it's the Module's internal thing. But the removal of this checkpoint should be persisted to DB along with the persist of latest batch of storage items, these two operations should be represented as an atomic DB modification.

Reply:
> so removing the checkpoint inside the statesync module seems okay.

Yes, that's true, let's keep it this way.